File: cpp/anagram/anagram.cpp
#include "anagram.h"
#include <algorithm>
#include <boost/algorithm/string.hpp>

vector<string> anagram::anagram::matches(initializer_list<string> list) {
    vector<string> v(list);
    vector<string> r;
    for (const string & s : v) {
        if (isAnagram(s)) {
            r.push_back(s);
        }
    }
    return r;
}

bool anagram::anagram::isAnagram(string s) {
    string a = this-> a;
    boost::algorithm::to_lower(s);
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    if (s != a && s.length() == a.length()) {
        for (const char &c : a) {
            int pos = s.find(c);
            if (pos < 0) {
                return false;
            }
            s.erase(s.begin() + pos);
        }
        return true;
    }
    return false;
}

anagram::anagram::anagram(string a) {
    boost::algorithm::to_lower(a);
    this->a = a;
}

File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where
import Data.List (sort)
import Data.Char (toLower)

anagramsFor :: String -> [String] -> [String]
anagramsFor s = filter (isAnagram s)  
   
isAnagram :: String -> String -> Bool
isAnagram a b = x /= y && sort x == sort y
    where
    x = map toLower a
    y = map toLower b
    
